vcf.c

```diff
@@ -4448,8 +4448,10 @@ int bcf_update_alleles(const bcf_hdr_t *hdr, bcf1_t *line, const char **alleles,
 
     // If the supplied alleles are not pointers to line->d.als, the existing block can be reused.
     int i;
+    uintptr_t als_start = (uintptr_t) line->d.als;
+    uintptr_t als_end = line->d.als ? als_start + line->d.m_als : als_start;
     for (i=0; i<nals; i++)
-        if ( alleles[i]>=line->d.als && alleles[i]<line->d.als+line->d.m_als ) break;
+        if ( (uintptr_t) alleles[i] >= als_start && (uintptr_t) alleles[i] < als_end ) break;
```

Comment:
Do you have an example of this?

I tried clang 10 with -fsanitize=undefined and this line as-is in develop, on the entire VCF test corpus (hts-specs#491) and cannot trigger this.

I find the rewrite harder to understand, so wonder if there is an issue whether it's simpler fixed.  Eg:

`if (line->d.als && alleles[i]>=line->d.als && alleles[i]<line->d.als+line->d.m_als ) break;`

Replacing the one line with 3 lines using `uintptr_t` really doesn't add to the clarity of what it's trying to do.

Reply:
The existing code relies on undefined behaviour, which puts it at risk of being [optimised in an unintended way](https://devblogs.microsoft.com/oldnewthing/20170927-00/?p=97095).  Using `uintptr_t` is also strictly undefined, but I suspect it's less likely that half the test would disappear because the optimiser thinks it shouldn't happen.

I do have an alternative fix that removes the undefined behaviour completely, but it's a far bigger change and I thought it best to go for a simple one for now.

Reply:
That still doesn't explain why the far simpler fix isn't sufficient, as suggested.  It avoids the use of uintptr_t too.

I did manage to reproduce this in the end with `./test/test-vcf-api test/test-vcf-api.out` and can confirm adding `line->d.als &&` to the if statement is all that is needed.  It seems far clearer too as it's obvious what it's doing: only do the bounds check if d.als is non-NULL.

Looking at the test code, alleles may be NULL, but nals is also 0 then so the loop isn't entered.  The other scenarios are (with alleles as non-NULL) that alleles points to within `line->d.als` in which case the `&&` claus above ensures we're valid (barring strict behaviour on pointer comparisons), or `line->d.als` is NULL (the bug we're trying to fix) in which case we break without pointer comparison as an indication that alleles is not an internal part of the existing line.

Reply:
I'm trying to fix both the bit of undefined behaviour that UBSAN complained about (adding zero to NULL) and the bit it didn't (comparing pointers to memory regions that are not part of the same array).